tests/test_lzip_decoder.c

```diff
@@ -0,0 +1,454 @@
+	uint8_t output_buffer[sizeof(expected_result) - 1];
+
+	strm.avail_in = file_size;
+	strm.next_in = data;
+	strm.avail_out = sizeof(output_buffer);
+	strm.next_out = output_buffer;
+
+	assert_lzma_ret(lzma_code(&strm, LZMA_FINISH), LZMA_DATA_ERROR);
+
+	// Test ignoring the checksum value - should decode successfully
+	assert_lzma_ret(lzma_lzip_decoder(&strm, MEMLIMIT,
+			LZMA_CONCATENATED | LZMA_IGNORE_CHECK), LZMA_OK);
+
+	strm.avail_in = file_size;
+	strm.next_in = data;
+	strm.avail_out = sizeof(output_buffer);
+	strm.next_out = output_buffer;
+
+	assert_lzma_ret(lzma_code(&strm, LZMA_FINISH), LZMA_STREAM_END);
+
+	assert_uint_eq(strm.total_in, file_size);
+	assert_uint_eq(strm.total_out, sizeof(expected_result) - 1);
+	assert_array_eq(output_buffer, expected_result, strm.total_out);
+
+	// Test tell check
+	assert_lzma_ret(lzma_lzip_decoder(&strm, MEMLIMIT,
+			LZMA_CONCATENATED | LZMA_TELL_ANY_CHECK), LZMA_OK);
+
+	strm.avail_in = file_size;
+	strm.next_in = data;
+	strm.avail_out = sizeof(output_buffer);
+	strm.next_out = output_buffer;
+
+	assert_lzma_ret(lzma_code(&strm, LZMA_FINISH), LZMA_GET_CHECK);
+	assert_uint_eq(lzma_get_check(&strm), LZMA_CHECK_CRC32);
+	assert_lzma_ret(lzma_code(&strm, LZMA_FINISH), LZMA_DATA_ERROR);
+	lzma_end(&strm);
+}
+
+
+static void
+test_invalid_magic_bytes(void) {
+	uint8_t lzip_id_string[] = { 0x4C, 0x5A, 0x49, 0x50 };
+	lzma_stream strm = LZMA_STREAM_INIT;
+
+	for (uint32_t i = 0; i < ARRAY_SIZE(lzip_id_string); i++) {
+		// Corrupt magic bytes
+		lzip_id_string[i] ^= 1;
+		uint8_t output_buffer[DECODE_CHUNK_SIZE];
+
+		assert_lzma_ret(lzma_lzip_decoder(&strm, MEMLIMIT, 0),
+				LZMA_OK);
+
+		strm.next_in = lzip_id_string;
+		strm.avail_in = sizeof(lzip_id_string);
+		strm.next_out = output_buffer;
+		strm.avail_out = DECODE_CHUNK_SIZE;
+
+		assert_lzma_ret(lzma_code(&strm, LZMA_RUN),
+				LZMA_FORMAT_ERROR);
+
+		// Reset magic bytes
+		lzip_id_string[i] ^= 1;
+	}
+
+	lzma_end(&strm);
+}
+
+
+static void
+test_invalid_version(void)
+{
+	// The file contains a version number that is not 0 or 1,
+	// so it should cause an error
+	decode_expect_error("files/unsupported-1-v234.lz",
+			LZMA_OPTIONS_ERROR);
+}
+
+
+static void
+test_invalid_dictionary_size(void) {
+	// First file has too small dictionary size field
+	decode_expect_error("files/bad-1-v1-dict-1.lz", LZMA_DATA_ERROR);
+	
+	// Second file has too large dictionary size field
+	decode_expect_error("files/bad-1-v1-dict-2.lz", LZMA_DATA_ERROR);
+}
+
+
+static void
+test_invalid_uncomp_size(void) {
+	// Test invalid v0 lzip file uncomp size
+	decode_expect_error("files/bad-1-v0-uncomp-size.lz",
+			LZMA_DATA_ERROR);
+	
+	// Test invalid v1 lzip file uncomp size
+	decode_expect_error("files/bad-1-v1-uncomp-size.lz",
+			LZMA_DATA_ERROR);
+}
+
+
+static void
+test_invalid_member_size(void) {
+	decode_expect_error("files/bad-1-v1-member-size.lz",
+			LZMA_DATA_ERROR);
+}
+
+
+static void
+test_invalid_memlimit(void) {
```

Comment:
Perhaps this test is useful but for portability it could be simplified so that it doesn't check the output content or even size, the CRC32 will catch some errors anyway. That is, just focus on the memlimit testing only.

Reply:
I think there is value in showing that if we have too low memlimit, we can up it and continue decoding. This seems like a common use case that we need to assert works

Reply:
That's exactly what I tried to say.

Reply:
I would rather check the CRC on the output to make this a more self contained test. What if the memlimit code somehow disables CRC checking and corrupts the output? Reading the lzip_decoder.c, its obvious this isn't the case, but the test doesn't know that. Its very few lines of code to check the output and makes the test more complete